libflux/src/core/build.rs

```diff
@@ -0,0 +1,38 @@
+// build.rs
+
+use std::env;
+use std::fs::copy;
+use std::path::PathBuf;
+
+// Bring in a dependency on an externally maintained `cc` package which manages
+// invoking the C compiler.
+extern crate cc;
+
+fn main() {
+    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
+
+    // The bindgen::Builder is the main entry point
+    // to bindgen, and lets you build up options for
+    // the resulting bindings.
+    let bindings = bindgen::Builder::default()
+        // The input header we would like to generate
+        // bindings for.
+        .header("scanner/scanner.h")
+        // Finish the builder and generate the bindings.
+        .generate()
+        // Unwrap the Result and panic on failure.
+        .expect("Unable to generate bindings");
+
+    // Write the bindings to the $OUT_DIR/bindings.rs file.
+    bindings
+        .write_to_file(out_path.join("bindings.rs"))
+        .expect("Couldn't write bindings!");
+
+    copy("../../scanner.c", out_path.join("scanner.c")).expect("Could not copy scanner.c");
+
+    // Compile generated scanner
+    cc::Build::new()
+        .include("scanner")
+        .file(out_path.join("scanner.c"))
+        .compile("scanner");
+}
```

Comment:
I honestly don't know much about Bindgen or why this file is necessary. Do you mind explaining a little bit?

Reply:
Why was this separated from the other code in the other `build.rs` file? Again just asking out of ignorance.

Reply:
That's the code that generates the bindings for the scanner. The scanner is written in ragel, which is compiled to C. The rust parser needs to call into that and this is the part that makes  the little bit of code that enables the call from Rust to C. I don't believe I changed that part. Unfortunately the patch makes it look like big changes.

If you haven't seen it already, have a look at the rust reference on build scripts. I just learned about it myself. Basically the above code is compiled runs just before cargo tries to build the module, so it can do things like generate bindings and other preparation for building.

https://doc.rust-lang.org/cargo/reference/build-scripts.html